sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -416,6 +416,29 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
         // these branches can be pruned away
         val (h, t) = branches.span(_._1 != TrueLiteral)
         CaseWhen( h :+ t.head, None)
+
+      case e @ CaseWhen(branches, _) =>
+        val newBranches = branches.foldLeft(List[(Expression, Expression)]()) {
+          case (newBranches, branch) =>
+            if (newBranches.exists(_._1.semanticEquals(branch._1))) {
+              // If a condition in a branch is previously seen, this branch can be pruned.
+              // TODO: In fact, if a condition is a sub-condition of the previous one,
+              // TODO: it can be pruned. This is less strict and can be implemented
+              // TODO: by decomposing seen conditions.
+              newBranches
```

Comment:
This seems good as the branch is useless. Removing it should simplify code and query plan.